README.md
trab-siad


CHANGELOG.md
# Changelog

## [Entrega 5] - 2020-06-29

### Adicionado

- Código Python *first-analisis.py*.

## [Entrega 4] - 2020-06-21

### Adicionado

- Adicionado nova tabela *indice_isolamento_social* no DataWarehouse.
- Nova planilha de dados sobre isolamento social.
- Adicionado novo script source-social-isolation-index.sql no qual injeta os dados no DW.

### Alterado

- Alterado o código source.sql adicionando mais datas no DW.
- Alterado o código coronavirus-structure.sql adicionando o código de criação da tabela *indice_isolamento_social*.
- Alterado o código source-covid-world.sql atualizando os dados do covid19 no mundo.
- Alterada a modelagem do DW.

## [Entrega 3] - 2020-06-12

### Corrigido

- Correção de escrita no cronograma de entregas.

### Alterado

- Datas do cronograma de entregas alteradas.

## [Entrega 2] - 2020-05-29

### Adicionado

- Scripts SQL de inserção de dados no DW para todas as tabelas com exceção da tabela *covid19-mundo*.

### Corrigido

- Correção do script de estrutura do DW.

## [Entrega 1] - 2020-05-20

### Adicionado

- Estrutura do Banco de Dados no SGBD PostgreSQL.
- Script SQL da estrutura do Banco de Dados.
- Novas tarefas adicionadas no cronograma de entregas parciais.

### Alterado

- Nome de atributos alterados no modelo do Data Warehouse.
- Alteração nas datas do cronograma de entregas parciais.

### Corrigido

- Correção de atributos e ligações nas tabelas do modelo do Data Warehouse.

## [Entrega 0] - 2020-05-08

### Adicionado

- Objetivo do projeto.
- Perguntas a serem respondidas.
- Fontes de dados.
- Base de dados.
- Modelagem do Data Warehouse.
- Cronograma de entregas.
- Script Pentaho para tratamento de dados.


scripts/Python/Scripts/first-analisis.py
""" Nome - RGA:
    Celso Antonio - 2014.1907.003-7
    Felipe S. Lopes - 2016.1907.032-4
"""

import psycopg2
import pandas as pd
import matplotlib.pyplot as plt
from datetime import datetime


def countries_graphs(country):
    """ Gera os gráficos para o país determinado.
        :param country: o país no qual terá um gráfico gerado.
    """
    df = pd.DataFrame()
    beds_number = list()

    try:
        # Query de busca de dados e montagem do data frame para análise.
        connection = psycopg2.connect(user="postgres",
                                      password="BodocoS0",
                                      host="localhost",
                                      port="5432",
                                      database="dw_covid19")

        cursor = connection.cursor()

        postgreSQL_select_query = "SELECT l.total_leitos, l.pais_iso " \
                                  "FROM coronavirus.leitos l " \
                                  "WHERE l.pais_iso = \'" + country + "\'"

        cursor.execute(postgreSQL_select_query)
        record = cursor.fetchall()

        for row in record:
            beds_number.append(row[0])

        print(beds_number)

        # Consultando mortes e casos em todos os dias por país.
        postgreSQL_select_query = "SELECT m.mortes, m.casos, t.dia, t.mes, t.ano, m.pais_cod " \
                                  "FROM coronavirus.covid19_mundo m JOIN coronavirus.tempo t ON data_reg = t.id " \
                                  "WHERE m.pais_cod = \'" + country + "\' " \
                                  "ORDER BY (ano, mes, dia)"


        cursor.execute(postgreSQL_select_query)
        record = cursor.fetchall()

        cases = list()
        deaths = list()
        time = list()
        covid_countries = list()

        for row in record:
            deaths.append(row[0])
            cases.append(row[1])
            date = str(row[2]) + "/" + str(row[3]) + "/" + str(row[4])
            date = datetime.strptime(date, '%d/%m/%Y')
            time.append(date)
            covid_countries.append(row[5])

        df['Mortes'] = deaths
        df['Casos'] = cases
        df['Data'] = time
        df['Pais'] = covid_countries
        df['Mortes Acumuladas'] = df.groupby('Pais')['Mortes'].cumsum()

    except (Exception, psycopg2.Error) as error:
        print("Error while connecting to PostgreSQL", error)
    finally:
        # closing database connection.
        if (connection):
            cursor.close()
            connection.close()
            print("PostgreSQL connection is closed")

    df2 = df.set_index('Data')
    df3 = df2.pivot(columns='Pais')

    ax = df3.plot()
    ax.axhline(beds_number[0], linewidth=1, color='r', label='Capacidade Máxima de Leitos')
    ax.legend()
    #ax.set_yscale('log')

    #plt.ylabel('Leitos')

    print(df['Data'])

    plt.show()


countries_graphs('BRA')
countries_graphs('ITA')
countries_graphs('USA')
countries_graphs('ESP')


scripts/Python/Scripts/third-analisis.py
""" Nome - RGA:
    Celso Antonio - 2014.1907.003-7
    Felipe S. Lopes - 2016.1907.032-4
"""

import psycopg2
import pandas as pd
import matplotlib.pyplot as plt
from datetime import datetime

df = pd.DataFrame()

def cities_graph(city):
    try:
        # Query de busca de dados e montagem do data frame para análise.
        connection = psycopg2.connect(user="postgres",
                                      password="BodocoS0",
                                      host="localhost",
                                      port="5432",
                                      database="dw_covid19")

        cursor = connection.cursor()
        postgreSQL_select_query = "SELECT ci.nome, cc.casos, tp.dia, tp.mes, tp.ano " \
                                  "FROM coronavirus.casos_por_cidade cc " \
                                  "JOIN coronavirus.tempo tp ON cc.id_data = tp.id " \
                                  "JOIN coronavirus.cidade ci ON ci.id = cc.cidade WHERE ci.nome = \'" + city + "\' " \
                                  "ORDER BY (tp.ano, tp.mes, tp.dia)"

        cursor.execute(postgreSQL_select_query)
        record = cursor.fetchall()

        city_row = list()
        cases = list()
        date = list()

        for row in record:
            city_row.append(row[0])
            cases.append(row[1])
            time = str(row[2]) + "/" + str(row[3]) + "/" + str(row[4])
            time = datetime.strptime(time, '%d/%m/%Y')
            date.append(time)

        df['Cidade'] = city_row
        df['Casos'] = cases
        df['Data'] = date

        postgreSQL_select_query = "SELECT ci.nome, tr.temp_celsius, tr2.temp_celsius, " \
                                  "tp.dia, tp.mes, tp.ano FROM coronavirus.temperatura_cidade tc " \
                                  "JOIN coronavirus.temperatura tr ON tc.temp_max = tr.id " \
                                  "JOIN coronavirus.temperatura tr2 ON tc.temp_min = tr2.id " \
                                  "JOIN coronavirus.cidade ci ON tc.cidade = ci.id " \
                                  "JOIN coronavirus.tempo tp ON tc.id_data = tp.id " \
                                  "WHERE ci.nome = \'" + city + "\' " \
                                  "ORDER BY (tp.ano, tp.mes, tp.dia) "

        cursor.execute(postgreSQL_select_query)
        record = cursor.fetchall()

        max_temp = list()
        min_temp = list()

        for row in record:
            max_temp.append(row[1])
            min_temp.append(row[2])

        df['Temperatura Máxima'] = max_temp
        df['Temperatura Mínima'] = min_temp
        print(df)


    except (Exception, psycopg2.Error) as error:
        print("Error while connecting to PostgreSQL", error)
    finally:
        # closing database connection.
        if (connection):
            cursor.close()
            connection.close()
            print("PostgreSQL connection is closed")

    df2 = df.set_index('Data')
    df3 = df2.pivot(columns='Cidade')

    df3.plot()
    plt.show()

cities_graph('São Paulo')


scripts/Python/Scripts/machine-learning.py
# Felipe Salles Lopes
# Celso Antonio

import psycopg2
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler
from sklearn.neural_network import MLPRegressor
from sklearn.model_selection import cross_val_predict
from sklearn import metrics
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVR

try:
    df = pd.DataFrame()

    connection = psycopg2.connect(user="postgres",
                                  password="BodocoS0",
                                  host="localhost",
                                  port="5432",
                                  database="dw_covid19")

    cursor = connection.cursor()

    postgreSQL_select_query = "SELECT tempo.dia, tempo.mes, tempo.ano, indice.taxa_isolamento " \
                              "FROM coronavirus.tempo as tempo, coronavirus.indice_isolamento_social as indice  " \
                              "WHERE tempo.id = indice.data_reg "

    cursor.execute(postgreSQL_select_query)
    record = cursor.fetchall()

    time = list()
    rate = list()
    for row in record:
        date = str(row[0]) + "/" + str(row[1]) + "/" + str(row[2])
        time.append(date)
        rate.append(row[3])

    df["Data"] = time
    df["Isolamento %"] = rate

    postgreSQL_select_query = "SELECT mundo.casos, mundo.mortes, mundo.pais_cod, tempo.dia, tempo.mes, tempo.ano " \
                              "FROM coronavirus.covid19_mundo as mundo " \
                              "JOIN coronavirus.tempo as tempo ON mundo.data_reg = tempo.id " \
                              "JOIN coronavirus.indice_isolamento_social as indice ON indice.data_reg = tempo.id " \
                              "WHERE pais_cod = 'BRA'"

    cursor.execute(postgreSQL_select_query)
    record = cursor.fetchall()

    cases = list()
    deaths = list()
    for row in record:
        cases.append(row[0])
        deaths.append(row[1])

    df["Casos"] = cases
    df["Mortes"] = deaths

    cases_norm = df["Casos"].values.copy()
    cases_norm.shape = (len(cases_norm), 1)

    scaler = StandardScaler()
    cases_norm = scaler.fit_transform(cases_norm)
    df["casos_normalizados"] = cases_norm

    window = pd.DataFrame()
    window_size = 3

    for i in range(0, window_size, +1):
        window["E{}".format(i)] = df["casos_normalizados"].shift(-i)

    window = window.iloc[:-window_size]

    holder = window.values

    X = holder[:, :-1]
    Y = holder[:, -1]

    print(df)
    print(window)

    mlp = MLPRegressor(hidden_layer_sizes=(50, 100,), activation='logistic', max_iter=1000, solver='sgd', verbose=True)
    svr = SVR(C=1.0, epsilon=0.2, verbose=True, max_iter=1000)

    full_predict_mlp = cross_val_predict(mlp, X, Y, cv=2)
    full_predict_svr = cross_val_predict(svr, X, Y, cv=2)

    print('Mean Squared Error in MLP: %s' % (metrics.mean_squared_error(Y, full_predict_mlp)))
    print('Mean Squared Error in SVR: %s' % (metrics.mean_squared_error(Y, full_predict_svr)))

    print(full_predict_svr.shape)
    print(full_predict_mlp.shape)

    df['Predict_mlp'] = full_predict_mlp
    df['Predict_svr'] = full_predict_svr

    df.plot(y=['casos_normalizados', 'Predict_mlp', 'Predict_svr'], style=['-s', '--o'])
    plt.show()

except (Exception, psycopg2.Error) as error:
    print("Error while connecting to PostgreSQL", error)
finally:
    # closing database connection.
    if (connection):
        cursor.close()
        connection.close()
        print("PostgreSQL connection is closed")

scripts/Python/Scripts/second-analisis.py
# Felipe Salles Lopes
# Celso Antonio

import psycopg2
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler
from sklearn.neural_network import MLPRegressor
from sklearn.model_selection import cross_val_predict
from sklearn import metrics
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVR


df = pd.DataFrame()
try:
    # Query de busca de dados e montagem do data frame para análise.
    connection = psycopg2.connect(user="postgres",
                                  password="BodocoS0",
                                  host="localhost",
                                  port="5432",
                                  database="dw_covid19")

    cursor = connection.cursor()

    # Obtenção dos dias nos quais têm a taxa de isolamento disponível.
    postgreSQL_select_query = "SELECT tempo.dia, tempo.mes, tempo.ano, indice.taxa_isolamento " \
                              "FROM coronavirus.tempo as tempo, coronavirus.indice_isolamento_social as indice  " \
                              "WHERE tempo.id = indice.data_reg "

    cursor.execute(postgreSQL_select_query)
    record = cursor.fetchall()

    time = list()
    rate = list()

    for row in record:
        # Montando a data no formato de String
        date = str(row[0]) + "/" + str(row[1]) + "/" + str(row[2])
        time.append(date)
        rate.append(row[3])

    df["Data"] = time
    df["Isolamento %"] = rate

    # Obtenção dos casos e mortes no Brasil para os dias selecionados na consulta anterior.
    postgreSQL_select_query = "SELECT mundo.casos, mundo.mortes, mundo.pais_cod, tempo.dia, tempo.mes, tempo.ano " \
                              "FROM coronavirus.covid19_mundo as mundo " \
                              "JOIN coronavirus.tempo as tempo ON mundo.data_reg = tempo.id " \
                              "JOIN coronavirus.indice_isolamento_social as indice ON indice.data_reg = tempo.id " \
                              "WHERE pais_cod = 'BRA'"

    cursor.execute(postgreSQL_select_query)
    record = cursor.fetchall()

    cases = list()
    deaths = list()
    for row in record:
        cases.append(row[0])
        deaths.append(row[1])

    df["Casos"] = cases
    df["Mortes"] = deaths

except (Exception, psycopg2.Error) as error:
    print("Error while connecting to PostgreSQL", error)
finally:
    # closing database connection.
    if (connection):
        cursor.close()
        connection.close()
        print("PostgreSQL connection is closed")

# Normalização dos casos
cases_norm = df["Casos"].values.copy()
cases_norm.shape = (len(cases_norm), 1)

scaler = StandardScaler()
cases_norm = scaler.fit_transform(cases_norm)
df["casos_normalizados"] = cases_norm

# Predição em forma de janela, olhando o tamanho da janela casos anteriores
window = pd.DataFrame()
window_size = 3

for i in range(0, window_size, +1):
    window["E{}".format(i)] = df["casos_normalizados"].shift(-i)

window = window.iloc[:-window_size]

holder = window.values

X = holder[:, :-1]
Y = holder[:, -1]

print(df)
print(window)

# Modelos de preditores
mlp = MLPRegressor(hidden_layer_sizes=(50, 100,), activation='logistic', max_iter=1000, solver='sgd', verbose=True)
svr = SVR(C=1.0, epsilon=0.2, verbose=True, max_iter=1000)

# Avaliação de preditores (com mais dados aumentar o cross validation)
full_predict_mlp = cross_val_predict(mlp, X, Y, cv=2)
full_predict_svr = cross_val_predict(svr, X, Y, cv=2)

print('Mean Squared Error in MLP: %s' % (metrics.mean_squared_error(Y, full_predict_mlp)))
print('Mean Squared Error in SVR: %s' % (metrics.mean_squared_error(Y, full_predict_svr)))

print(full_predict_svr.shape)
print(full_predict_mlp.shape)

# Tratamento para plotar o gráfico
values_to_add = list()
for i in range(0, window_size):
    values_to_add.append(float('NaN'))

full_predict_svr = np.insert(full_predict_svr, 0, values_to_add)
full_predict_svr.shape = (len(full_predict_svr), 1)


full_predict_mlp = np.insert(full_predict_mlp, 0, values_to_add)
full_predict_mlp.shape = (len(full_predict_mlp), 1)

print(full_predict_svr.shape)
print(full_predict_mlp.shape)

df['Predict_mlp'] = full_predict_mlp
df['Predict_svr'] = full_predict_svr

df.plot(y=['casos_normalizados', 'Predict_mlp', 'Predict_svr'], style=['-s', '--o'])
plt.show()